ltree_io.c

```diff
@@ -90,35 +98,36 @@ ltree_in(PG_FUNCTION_ARGS)
 	while (*ptr)
 	{
 		charlen = pg_mblen(ptr);
+		delimiter = false;
 
 		if (state == LTPRS_WAITNAME)
 		{
 			lptr->start = ptr;
 			lptr->wlen = 0;
 
 			/* handle corner case when label starts with a delimiter character */
-			state = (charlen == 1 && t_iseq(ptr, NODE_DELIMITER_CHAR)) ?
-				LTPRS_WAITDELIMEND : LTPRS_WAITDELIM;
+			if (is_delimiter(ptr, charlen))
+				ereport(ERROR,
+							(errcode(ERRCODE_SYNTAX_ERROR),
+							 errmsg("syntax error"),
+							 errdetail("Unexpected delimeter in position %d",
+									   pos)));
+			state = LTPRS_WAITDELIM;
 		}
-		else if (state == LTPRS_WAITDELIMEND)
+		else if (state == LTPRS_WAITDELIM)
 		{
-			if (charlen == 1 && t_iseq(ptr, NODE_DELIMITER_CHAR))
+			if (is_delimiter(ptr, charlen))
 			{
-				lptr->len = ptr - lptr->start - 1;
-				lptr->wlen--;
+				delimiter = true;
```

Comment:
since we call `is_delimiter` in both cases, it would be shorter to do something like this:
```c
delimiter = is_delimiter(ptr, charlen);
if (state == LTPRS_WAITNAME)
{
    if (delimiter) { ... }
}
else if (state == LTPRS_WAITDELIM)
{
    if (delimiter) { ... }
}
```

Reply:
Agree. I simplified the code to avoid multiple `is_delimiter` calls.